keps/sig-storage/3333-reconcile-default-storage-class/README.md

```diff
@@ -0,0 +1,656 @@
+
+1. It’s hard to mark a different SC as the default one. Cluster admin can
+   choose between two bad solutions:
+
+    1. Cluster has two default SCs for a short time, i.e. admin marks the new
+       default SC as default and then marks the old default SC as non-default. When
+       there are two default SCs in a cluster, the PVC admission plugin refuses to
+       accept new PVCs with `pvc.spec.storageClassName = nil`. Hence, cluster users
+       may get errors when creating PVCs at the wrong time. They must know it’s a
+       transient error and manually retry later.
+
+    2. Cluster has no default SC for a short time, i.e. admin marks the old
+       default SC as non-default and then marks the new default SC as default.
+       Since there is no default SC for some time, PVCs with
+       `pvc.spec.storageClassName = nil` created during this time will not get any
+       SC and are Pending forever. Users must be smart enough to delete the PVC and
+       re-create it.
+
+2. When users want to change the default SC parameters, they must delete the SC
+   and re-create it, Kubernetes API does not allow change in the SC. So there is no
+   default SC for some time and second case above applies here too.
+
+3. Defined ordering during cluster installation. Kubernetes cluster installation
+   tools must be currently smart enough to create a default SC before starting
+   anything that may create PVCs that need it. If such a tool supports multiple
+   cloud providers, storage backends and addons that require storage (such an image
+   registry), it may be quite complicated to do the ordering right.
+
+4. If a default storage class does not exist, PVCs with
+   `pvc.spec.storageClassName=nil` are bound to PVs with
+   `pv.spec.storageClassName=""`. This can be confusing to users, the PVC is bound
+   to different PVs or dynamically provisioned, depending on if the default SC
+   exists or not at the time when the PVC is created.
+
+### Goals
+
+* Loosen ordering requirements between PVCs and default SC creation.
+* Improve user experience with `pvc.spec.storageClassName=nil`.
+
+### Non-Goals
+
+* Introduce new API for default SCs.
+
+## Proposal
+
+Right now, the default SC is applied to PVCs in an admission plugin, i.e. only
+during PVC creation. Later on, PV controller will bind PVCs with
+`pvc.spec.storageClassName=nil` to PVs with `pv.spec.storageClassName=""`,
+if such a PV exists.
+
+We want to change the existing `pvc.spec.storageClassName=nil` behavior
+to always require a default SC. It would never bind to a PV with
+`pv.spec.storageClassName=""` and it would be `Pending` until a default SC
+is created. When a default SC is created, the PVC `pvc.spec.storageClassName`
+will be updated with the new default SC name in the PV controller.
+
+Any PVs with `pv.spec.storageClassName=""` will be able to bind only to a PVC
+with `pvc.spec.storageClassName=""`.
+
+This behavior should be simpler from the user perspective.
```

Comment:
Does this proposal include any changes to the PV admission plugin?  Please clarify.

Reply:
No. Adding a clarification that the behavior improvement does not need PV admission plugin changes.